### PR TITLE
Support using a single key as input for instantiating HDWalletPro…

### PR DESCRIPTION
**Pull request by eggplantzzz**
…vider

### CODE REVIEW

**Review by richardpringle — APPROVED** [review]

**Comment by richardpringle** [comment]
@gnidan, I'll let you take a look before merging (you might be able to merge yourself too!)

**Comment by eggplantzzz** [comment]
I'll ping @gnidan to make sure he gets a chance to look at these changes.

File: packages/hdwallet-provider/src/constructor/Constructor.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/Constructor.ts]: Note that doing this removes type-safety, since both may be optional and the type-checker would consider it a valid instance of the interface.

In this case it doesn't matter, since it the code that generates these objects isn't capable of generating an object with neither of those properties, but still... I went with this approach in the first place with the goal to be as explicit as possible ;)

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts]: This type is misleading if it doesn't mention `PrivateKey` as a third option... seems like there's room to come up with a name for this.

I propose `credentials: Credentials`.

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts]: This is going to break type safety!!!

This function tells TypeScript, "I will return a boolean value, and, based on that boolean, you will whether the type you asked about is a `PrivateKey[]`"... by returning `true` for strings, you break that promise, and TS is left none-the-wiser.

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts]: This should stay the same as the argument to `isMnemonicPhrase` (well, both should have all 3 options, really)... the idea of these functions are to take something that can be one of many things and tell the type-checker how it knows when it's figured out which one it is.

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts]: Instead of this, it's more clear to have a third conditional for `isPrivateKey`.

Going with my suggestion, to define a `LegacyConstructor.Credentials` type:
```typescript
export type Credentials = MnemonicPhrase | PrivateKey | PrivateKey[];
```

Then we'd define three separate type predicates (see [docs about this](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates)):
- `isMnemonic`
- `isPrivateKey`
- `isPrivateKeys`

The idea is that each of these takes a value of any of the three allowed types and returns a boolean based on which one it is. This gets used both at runtime, for conditional logic, and at compile-time, to tell TS not to barf on `for (const privateKey of mnemonicPhraseOrPrivateKeys)` if you wrap it in `if (isPrivateKeys(mnemonicPhraseOrPrivateKeys)) { ... }`.

So then, in this function, we could just lean on having type predicates for each possibility, and move the ternary operator into a new `else if` that's at the same top-level.

File: packages/hdwallet-provider/src/index.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/index.ts:125]: This is a good check to include! (What prompted you to add it? I'm curious.)

File: packages/hdwallet-provider/test/provider.test.ts
Comment (gnidan) [inline packages/hdwallet-provider/test/provider.test.ts]: I think we should disallow the single-`privateKey` case in the new interface, while we still can.

My reasoning:

- It's very little cost for users to do `{ privateKeys: [privateKey] }` instead of `{ privateKey }`.
  - Naturally, `privateKeys` works perfectly well for both cases, so there's no negative impact except possible confusion.
  - We can remove risk of confusion by specifying "how to pass a single private key" in the README (although I'm not sure it's necessary).
- Including this option adds extra logic:
  - Something has to distinguish this case at run-time (like you've added to `getPrivateKeys`).
  - The type descriptions need to account for the extra possibility, and that gets into risky areas because TS forces you to do either `{ a } | { b }` or `{ a?, b? }`. The latter here is type unsafe; the former adds something else to deal with.

File: packages/hdwallet-provider/test/provider.test.ts
Comment (gnidan) [inline packages/hdwallet-provider/test/provider.test.ts:122]: Thanks for making tests for this!

**Review by gnidan — CHANGES_REQUESTED** [review]
@eggplantzzz got some revisions for you. I opened a PR [here](https://github.com/trufflesuite/truffle/pull/3340) because I figured it'd be easier to show what I mean by example.

Sorry for being unclear in my write-up of this remaining change to @richardpringle's PR. I appreciate your offering to take this change off my plate... I would've missed a few of the changes here and probably let this sit another month! 😮 🙇

File: packages/hdwallet-provider/test/provider.test.ts
Comment (eggplantzzz) [reply packages/hdwallet-provider/test/provider.test.ts]: I'm not opposed to not supporting `privateKey`. It seems kind of unnecessary as it is just a special case of `PrivateKey[]`.  But I always hear your voice in the back of my head saying "we should handle/accept as wide a variety of input as possible" haha.

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts]: Oops comment here is wrong, I'll fix
```suggestion
// check that the first argument is a list of private keys
```

File: packages/hdwallet-provider/src/constructor/getOptions.ts
Comment (gnidan) [inline packages/hdwallet-provider/src/constructor/getOptions.ts:22]:
```suggestion
// check that the first argument is a single private key (default case for invalid mnemonics)
const isPrivateKey = (
```

**Review by gnidan — CHANGES_REQUESTED** [review]
(needs some comment fixing)

**Review by gnidan — APPROVED** [review]
I think this is the last of it before the original PR is a-go!

Will wait for the tests to turn green then hit merge on this one

**Pull request merged**
